File: 2018/src/main/scala/advent/solutions/Day1.scala
package advent.solutions

import advent.util.Util

object Day1 {

  def part1(input: Array[String]): Int = input.map(_.toInt).sum

  def part2(input: Array[String]): Option[Int] = {
    Stream.continually(input.toStream.map(_.toInt)).flatten.foldLeft((Set[Int](), 0)) { case ((results, prev), x) =>
      val result = prev + x
      if (results(result)) return Some(result) else (results + result, result)
    }
    None
  }

  def main(args: Array[String]): Unit = {
    println("Part 1 result: " + part1(Util.readFileLines(args(0))))
    println("Part 2 result: " + part2(Util.readFileLines(args(1))).get)
  }

}

File: 2019/src/main/kotlin/hr/dodomix/advent/solutions/Day1.kt
package hr.dodomix.advent.solutions

import hr.dodomix.advent.util.Util

class Day1 {
    fun dayDirectory() = "day1"

    fun part1(input: List<String>): Int =
        input.map { it.toInt() }
            .map { it / 3 }
            .map { it - 2 }
            .sum()

    private fun calculateFuel(value: Int): Int =
        (value / 3 - 2)
            .takeIf { it > 0 }
            ?.let { it + calculateFuel(it) }
            ?: 0

    fun part2(input: List<String>): Int = input
        .map { it.toInt() }
        .map { calculateFuel(it) }
        .sum()
}